File: backend/src/firewall.rs
use std::process::Command;
use crate::config::Config;

pub struct FirewallManager {
    trusted_ips: Vec<String>,
}

impl FirewallManager {
    pub fn new() -> Self {
        Self {
            trusted_ips: vec!["127.0.0.1".to_string(), "::1".to_string()],
        }
    }

    fn is_safe_to_block(&self, ip: &str) -> bool {
        if self.trusted_ips.contains(&ip.to_string()) {
            return false;
        }
        if ip.starts_with("224.") || ip.starts_with("239.") || ip.ends_with(".255") {
            return false;
        }
        true
    }

    pub fn setup_gateway(&self, config: &Config) {
        if config.idps_deployment_mode != "GATEWAY" {
            return;
        }

        println!("FirewallManager: Setting up Gateway NAT routing: WAN={}, LAN={}", config.wan_interface, config.lan_interface);
        
        let _ = Command::new("sudo").args(&["sysctl", "-w", "net.ipv4.ip_forward=1"]).output();
        
        let _ = Command::new("sudo").args(&["iptables", "-t", "nat", "-A", "POSTROUTING", "-o", &config.wan_interface, "-j", "MASQUERADE"]).output();
        let _ = Command::new("sudo").args(&["iptables", "-A", "FORWARD", "-i", &config.wan_interface, "-o", &config.lan_interface, "-m", "state", "--state", "RELATED,ESTABLISHED", "-j", "ACCEPT"]).output();
        let _ = Command::new("sudo").args(&["iptables", "-A", "FORWARD", "-i", &config.lan_interface, "-o", &config.wan_interface, "-j", "ACCEPT"]).output();
        
        println!("FirewallManager: Gateway NAT routing configured successfully.");
    }

    pub fn teardown_gateway(&self, config: &Config) {
        if config.idps_deployment_mode != "GATEWAY" {
            return;
        }

        println!("FirewallManager: Tearing down Gateway NAT routing: WAN={}, LAN={}", config.wan_interface, config.lan_interface);
        
        let _ = Command::new("sudo").args(&["iptables", "-t", "nat", "-D", "POSTROUTING", "-o", &config.wan_interface, "-j", "MASQUERADE"]).output();
        let _ = Command::new("sudo").args(&["iptables", "-D", "FORWARD", "-i", &config.wan_interface, "-o", &config.lan_interface, "-m", "state", "--state", "RELATED,ESTABLISHED", "-j", "ACCEPT"]).output();
        let _ = Command::new("sudo").args(&["iptables", "-D", "FORWARD", "-i", &config.lan_interface, "-o", &config.wan_interface, "-j", "ACCEPT"]).output();
        
        println!("FirewallManager: Gateway NAT routing rules removed.");
    }

    pub fn block_ip(&self, ip: &str, config: &Config) -> bool {
        if !self.is_safe_to_block(ip) {
            println!("FirewallManager: Refused to block trusted/unsafe IP {}", ip);
            return false;
        }

        // INPUT chain
        let mut success = true;
        let output = Command::new("sudo")
            .args(&["iptables", "-A", "INPUT", "-s", ip, "-j", "DROP"])
            .output();
        
        match output {
            Ok(out) if out.status.success() => println!("FirewallManager: Added Linux rule to block {} on INPUT", ip),
            _ => {
                println!("FirewallManager: Failed to block {} on INPUT", ip);
                success = false;
            }
        }

        // FORWARD chain if gateway
        if config.idps_deployment_mode == "GATEWAY" {
            let output_fwd = Command::new("sudo")
                .args(&["iptables", "-A", "FORWARD", "-s", ip, "-j", "DROP"])
                .output();
            match output_fwd {
                Ok(out) if out.status.success() => println!("FirewallManager: Added Linux rule to block {} on FORWARD", ip),
                _ => {
                    println!("FirewallManager: Failed to block {} on FORWARD", ip);
                    success = false;
                }
            }
        }
        success
    }

    pub fn unblock_ip(&self, ip: &str, config: &Config) -> bool {
        let mut success = true;
        let output = Command::new("sudo")
            .args(&["iptables", "-D", "INPUT", "-s", ip, "-j", "DROP"])
            .output();

        match output {
            Ok(out) if out.status.success() => println!("FirewallManager: Removed Linux rule for {} from INPUT", ip),
            _ => {
                println!("FirewallManager: Failed to unblock {} from INPUT", ip);
                success = false;
            }
        }

        if config.idps_deployment_mode == "GATEWAY" {
            let output_fwd = Command::new("sudo")
                .args(&["iptables", "-D", "FORWARD", "-s", ip, "-j", "DROP"])
                .output();
            match output_fwd {
                Ok(out) if out.status.success() => println!("FirewallManager: Removed Linux rule for {} from FORWARD", ip),
                _ => {
                    println!("FirewallManager: Failed to unblock {} from FORWARD", ip);
                    success = false;
                }
            }
        }
        success
    }
}

File: backend/firewall/firewall.go
package firewall

import (
	"fmt"
	"net"
	"os"
	"os/exec"
	"strings"

	"idps-backend/config"
)

type FirewallManager struct {
	trustedIPs        []string
	ipForwardOriginal string
}

func NewFirewallManager() *FirewallManager {
	return &FirewallManager{
		trustedIPs: []string{"127.0.0.1", "::1"},
	}
}

func (fm *FirewallManager) isSafeToBlock(ip string, cfg *config.Config) bool {
	cfg.Mu.RLock()
	gatewayIP := cfg.GatewayIP
	cfg.Mu.RUnlock()

	if gatewayIP != "" && ip == gatewayIP {
		return false
	}
	for _, trusted := range fm.trustedIPs {
		if ip == trusted {
			return false
		}
	}
	if strings.HasPrefix(ip, "224.") || strings.HasPrefix(ip, "239.") || strings.HasSuffix(ip, ".255") {
		return false
	}

	addrs, err := net.InterfaceAddrs()
	if err == nil {
		for _, addr := range addrs {
			var localIP net.IP
			switch v := addr.(type) {
			case *net.IPNet:
				localIP = v.IP
			case *net.IPAddr:
				localIP = v.IP
			}
			if localIP != nil && localIP.String() == ip {
				return false
			}
		}
	}
	return true
}

// iptablesBin returns "ip6tables" for IPv6 addresses, "iptables" for IPv4.
func iptablesBin(ip string) string {
	if strings.Contains(ip, ":") {
		return "ip6tables"
	}
	return "iptables"
}

func (fm *FirewallManager) runCommand(cfg *config.Config, name string, args ...string) error {
	cfg.Mu.RLock()
	dryRun := cfg.FirewallDryRun
	cfg.Mu.RUnlock()

	if dryRun {
		fmt.Printf("[DRY RUN] Would execute: %s %s\n", name, strings.Join(args, " "))
		return nil
	}
	cmd := exec.Command(name, args...)
	return cmd.Run()
}

func (fm *FirewallManager) ruleExists(cfg *config.Config, bin string, args ...string) bool {
	cfg.Mu.RLock()
	dryRun := cfg.FirewallDryRun
	cfg.Mu.RUnlock()

	if dryRun {
		return false
	}
	// Build a -C (check) command from the original args.
	// -A CHAIN ... → -C CHAIN ...
	// -I CHAIN pos ... → -C CHAIN ...  (strip the position number)
	var checkArgs []string
	if len(args) > 0 && args[0] == "-I" && len(args) > 2 {
		// args = ["-I", "CHAIN", "pos", rest...]
		// Skip args[0] ("-I") and args[2] (position number)
		checkArgs = append([]string{"-C", args[1]}, args[3:]...)
	} else {
		checkArgs = append([]string{"-C"}, args[1:]...)
	}
	cmd := exec.Command("sudo", append([]string{bin}, checkArgs...)...)
	return cmd.Run() == nil
}

func (fm *FirewallManager) ensureRule(cfg *config.Config, bin string, args ...string) error {
	if !fm.ruleExists(cfg, bin, args...) {
		return fm.runCommand(cfg, "sudo", append([]string{bin}, args...)...)
	}
	return nil
}

func (fm *FirewallManager) SetupGateway(cfg *config.Config) error {
	cfg.Mu.RLock()
	mode := cfg.IDPSDeploymentMode
	wan := cfg.WanInterface
	lan := cfg.LanInterface
	dryRun := cfg.FirewallDryRun
	cfg.Mu.RUnlock()

	if mode != "GATEWAY" && mode != "NETWORK" {
		return nil
	}

	fmt.Printf("FirewallManager: Setting up Gateway NAT routing: WAN=%s, LAN=%s\n", wan, lan)

	if wan == "" || lan == "" {
		return fmt.Errorf("WAN or LAN interface is not specified")
	}
	if wan == lan {
		return fmt.Errorf("WAN and LAN interfaces cannot be the same")
	}

	// Validate interfaces exist
	if _, err := net.InterfaceByName(wan); err != nil && !dryRun {
		return fmt.Errorf("WAN interface %s not found: %v", wan, err)
	}
	if _, err := net.InterfaceByName(lan); err != nil && !dryRun {
		return fmt.Errorf("LAN interface %s not found: %v", lan, err)
	}

	if !dryRun {
		b, err := os.ReadFile("/proc/sys/net/ipv4/ip_forward")
		if err == nil {
			fm.ipForwardOriginal = strings.TrimSpace(string(b))
		}
	}

	if err := fm.runCommand(cfg, "sudo", "sysctl", "-w", "net.ipv4.ip_forward=1"); err != nil {
		return fmt.Errorf("failed to enable ipv4 forwarding: %v", err)
	}

	if !dryRun {
		b, _ := os.ReadFile("/proc/sys/net/ipv4/ip_forward")
		if strings.TrimSpace(string(b)) != "1" {
			return fmt.Errorf("IPv4 forwarding is not enabled (read %s)", string(b))
		}
	}

	// NAT Rule
	err := fm.ensureRule(cfg, "iptables", "-t", "nat", "-A", "POSTROUTING", "-o", wan, "-m", "comment", "--comment", "IDPS-NAT", "-j", "MASQUERADE")
	if err != nil {
		return fmt.Errorf("failed to configure NAT: %v", err)
	}

	// FORWARD WAN -> LAN (established/related)
	err = fm.ensureRule(cfg, "iptables", "-A", "FORWARD", "-i", wan, "-o", lan, "-m", "conntrack", "--ctstate", "RELATED,ESTABLISHED", "-m", "comment", "--comment", "IDPS-FWD-IN", "-j", "ACCEPT")
	if err != nil {
		return fmt.Errorf("failed to configure WAN->LAN forwarding: %v", err)
	}

	// FORWARD LAN -> WAN
	err = fm.ensureRule(cfg, "iptables", "-A", "FORWARD", "-i", lan, "-o", wan, "-m", "comment", "--comment", "IDPS-FWD-OUT", "-j", "ACCEPT")
	if err != nil {
		return fmt.Errorf("failed to configure LAN->WAN forwarding: %v", err)
	}

	fmt.Println("FirewallManager: Gateway NAT routing configured successfully.")
	return nil
}

func (fm *FirewallManager) TeardownGateway(cfg *config.Config) {
	cfg.Mu.RLock()
	mode := cfg.IDPSDeploymentMode
	wan := cfg.WanInterface
	lan := cfg.LanInterface
	dryRun := cfg.FirewallDryRun
	cfg.Mu.RUnlock()

	if mode != "GATEWAY" && mode != "NETWORK" {
		return
	}

	fmt.Printf("FirewallManager: Tearing down Gateway NAT routing: WAN=%s, LAN=%s\n", wan, lan)

	_ = fm.runCommand(cfg, "sudo", "iptables", "-t", "nat", "-D", "POSTROUTING", "-o", wan, "-m", "comment", "--comment", "IDPS-NAT", "-j", "MASQUERADE")
	_ = fm.runCommand(cfg, "sudo", "iptables", "-D", "FORWARD", "-i", wan, "-o", lan, "-m", "conntrack", "--ctstate", "RELATED,ESTABLISHED", "-m", "comment", "--comment", "IDPS-FWD-IN", "-j", "ACCEPT")
	_ = fm.runCommand(cfg, "sudo", "iptables", "-D", "FORWARD", "-i", lan, "-o", wan, "-m", "comment", "--comment", "IDPS-FWD-OUT", "-j", "ACCEPT")

	if fm.ipForwardOriginal == "0" && !dryRun {
		_ = fm.runCommand(cfg, "sudo", "sysctl", "-w", "net.ipv4.ip_forward=0")
	}

	fmt.Println("FirewallManager: Gateway NAT routing rules removed.")
}

func (fm *FirewallManager) BlockIP(ip string, cfg *config.Config) bool {
	cfg.Mu.RLock()
	mode := cfg.IDPSDeploymentMode
	cfg.Mu.RUnlock()

	if mode != "HOST" && mode != "GATEWAY" && mode != "NETWORK" {
		return false
	}

	if !fm.isSafeToBlock(ip, cfg) {
		fmt.Printf("FirewallManager: Refused to block trusted/unsafe IP %s\n", ip)
		return false
	}

	bin := iptablesBin(ip)
	success := true

	if mode == "HOST" {
		args := []string{"-I", "INPUT", "1", "-s", ip, "-m", "comment", "--comment", "IDPS-BLOCK", "-j", "DROP"}
		if err := fm.ensureRule(cfg, bin, args...); err == nil {
			fmt.Printf("FirewallManager: Added %s rule to block %s on INPUT\n", bin, ip)
		} else {
			fmt.Printf("FirewallManager: Failed to block %s on INPUT: %v\n", ip, err)
			success = false
		}
	} else if mode == "GATEWAY" || mode == "NETWORK" {
		args := []string{"-I", "FORWARD", "1", "-s", ip, "-m", "comment", "--comment", "IDPS-BLOCK", "-j", "DROP"}
		if err := fm.ensureRule(cfg, bin, args...); err == nil {
			fmt.Printf("FirewallManager: Added %s rule to block %s on FORWARD\n", bin, ip)
		} else {
			fmt.Printf("FirewallManager: Failed to block %s on FORWARD: %v\n", ip, err)
			success = false
		}
	}

	return success
}

func (fm *FirewallManager) UnblockIP(ip string, cfg *config.Config) bool {
	cfg.Mu.RLock()
	mode := cfg.IDPSDeploymentMode
	cfg.Mu.RUnlock()

	if mode != "HOST" && mode != "GATEWAY" && mode != "NETWORK" {
		return false
	}

	bin := iptablesBin(ip)
	success := true

	if mode == "HOST" {
		if err := fm.runCommand(cfg, "sudo", bin, "-D", "INPUT", "-s", ip, "-m", "comment", "--comment", "IDPS-BLOCK", "-j", "DROP"); err == nil {
			fmt.Printf("FirewallManager: Removed %s rule for %s from INPUT\n", bin, ip)
		} else {
			fmt.Printf("FirewallManager: Failed to unblock %s from INPUT: %v\n", ip, err)
			success = false
		}
	} else if mode == "GATEWAY" || mode == "NETWORK" {
		if err := fm.runCommand(cfg, "sudo", bin, "-D", "FORWARD", "-s", ip, "-m", "comment", "--comment", "IDPS-BLOCK", "-j", "DROP"); err == nil {
			fmt.Printf("FirewallManager: Removed %s rule for %s from FORWARD\n", bin, ip)
		} else {
			fmt.Printf("FirewallManager: Failed to unblock %s from FORWARD: %v\n", ip, err)
			success = false
		}
	}

	return success
}
